### PR TITLE
*: batch write to temp file

### PR DESCRIPTION
**Pull request by YuJuncen**
Before, we split a batch of raft events and write them one-by-one to the temporary file. When many batches arrives, each batch would be write concurrently. However, for keeping the order of the log, we MUST add a lock for each temporary file.

图像

This may make too many batches keep in main memory: note that we only release the memory of a batch when it FULLY consumed. (Because we cannot fast `pop_front` on `Vec`s) Memory leaking may happen if there are hotspot of writing.

This PR did two things:
1. Make each batch hold the lock until it is fully consumed.

| Master | This PR |
| --- | --- |
| image | image |
|image|image|

2. Use `BufWrite` for reducing the invocations of system call.

| Master | This PR |
| --- | --- |
|PLuXiNgwC6|image|

### CODE REVIEW

File: components/br-stream/src/event_loader.rs [hunk 1]
```diff
@@ -190,20 +193,18 @@ where
         let mut event_loader = EventLoader::load_from(snap, start_ts, TimeStamp::max(), region)?;
         let mut stats = StatisticsSummary::default();
         loop {
-            let mut events = Vec::with_capacity(2048);
+            let mut events = ApplyEvents::with_capacity(1024, region.id);
             let stat = event_loader.scan_batch(1024, &mut events)?;
-            if events.is_empty() {
+            if events.len() == 0 {
```
Comment (kennytm) [inline components/br-stream/src/event_loader.rs:198]: wouldn't clippy have warned that you have implemented `len()` but not `is_empty()`? 🤔

Comment (YuJuncen) [reply components/br-stream/src/event_loader.rs:198]: Nope...
image

Comment (kennytm) [reply components/br-stream/src/event_loader.rs:198]: okay 🤔

**Review by kennytm — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**